### ISSUE TITLE
Support PyTorch-lightning DDP training

### ISSUE THREAD

**Issue by nzw0301** [optuna/optuna#2699]
## Motivation

PyTorch lightning's DDP training is an efficient way to train neural networks on multiple GPUs. Some of Optuna's functions do not work with DDP, for example, callback as reported in https://github.com/optuna/optuna-examples/issues/14.

## Description

I'm not familiar with pytorch-lightning, but let me mention a few related features:

- [`torchmetrics`](https://torchmetrics.readthedocs.io/en/latest/pages/lightning.html) to sync evaluation metric among GPUs
- Optuna's [`TorchDistributedTrial`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.TorchDistributedTrial.html) for pure pytorch's DDP training
